src/shared/particle_dynamics/solid_dynamics/thin_structure_dynamics.cpp

```diff
@@ -140,11 +140,30 @@ void ShellStressRelaxationFirstHalf::initialization(size_t index_i, Real dt)
                                             transformation_matrix_[index_i] * current_transformation_matrix.transpose();
 
         /** correct Almansi strain tensor according to plane stress problem. */
-        current_local_almansi_strain = getCorrectedAlmansiStrain(current_local_almansi_strain, nu_);
+        // current_local_almansi_strain = getCorrectedAlmansiStrain(current_local_almansi_strain, nu_);
+        current_local_almansi_strain(2, 2) = 0;
 
         /** correct out-plane numerical damping. */
-        Matd cauchy_stress = elastic_solid_.StressCauchy(current_local_almansi_strain, F_gaussian_point, index_i) + current_transformation_matrix * transformation_matrix_[index_i].transpose() * F_gaussian_point *
-                                                                                                                        elastic_solid_.NumericalDampingRightCauchy(F_gaussian_point, dF_gaussian_point_dt, numerical_damping_scaling_[index_i], index_i) * F_gaussian_point.transpose() * transformation_matrix_[index_i] * current_transformation_matrix.transpose() / F_gaussian_point.determinant();
+        double E = E0_; // Take the default Young's modulus of the material as the initial derivative.
+        Matd cauchy_stress = elastic_solid_.StressCauchy(current_local_almansi_strain, F_gaussian_point, index_i);
+        int it = 0;
+        int max_it = 20;
+        for (double s_next = 1.0; abs(s_next) > 1e-9 && it < max_it; ++it)
+        {
+            double e_prev = current_local_almansi_strain(2, 2);
+            double s_prev = cauchy_stress(2, 2);
+            double e_next = e_prev - s_prev / E;
+            current_local_almansi_strain(2, 2) = e_next;
+            cauchy_stress = elastic_solid_.StressCauchy(current_local_almansi_strain, F_gaussian_point, index_i);
+            s_next = cauchy_stress(2, 2);
+            E = (s_next - s_prev) / (e_next - e_prev);
+            // de = std::abs(e_next - e_prev);
+        }
+        if (it == max_it)
+            throw std::runtime_error("Secant algorithm enforcing `Szz = 0` could not converge.");
```

Comment:
Could this condition be too strict? I'm worried that it might not converge for non-linear models. Can we use relative error criteria and replace the runtime error with a warning message?

Reply:
Yes it might be, it either converges quite fast (3~8 iterations) or completely diverge usually. But when it diverges then the stress is NaN, effectively it has not been reaching maximum iterations. Ideally it should be relative, but relative to what. It could be the max/min absolute principal stress of the upper block, the absolute max/min of the upper block diagonal, the Frobenius norm of the upper block...

Let's continue the discussion on #125